### ISSUE TITLE
Founding KubeVirt Common Language

### ISSUE THREAD

**Issue by davidvossel** [kubevirt/kubevirt#305]
Inconsistency in our terminology is making the latest round of proposals difficult to process (at least for me). For example, I *think* I understand what's being proposed, but there is enough ambiguity that it makes me uncertain I'm really following the the full intent.

Lets use this issue to solidify a set of common terms we can all agree on.

I tried to get us started below. I'll update the glossary.md with the results. 

**VM Spec**: *<some text here>* ... something to do with libvirt domain represented in kubevirt objects.
**VirtualMachineConfig**: A cluster object that represents a static VM spec. Configs can be used to launch a VM Instance.
**VM Instance**: A running virtual machine
**VM**: A cluster object that represents a running VM Instance. This objectcContains the VM Spec used to launch the VM Instance.
**VMPreset**: A cluster object that ... *<insert txt here>*
**VM Spec Completion**: The action of filling in a sparse VM specification through VM Spec defaulting, VMPreset injection, and libvirt domain defaulting *<more text here>*
**VM Flavor**: *<insert text here>*
**VM Spec Defaulting**: The action of ... *<some text here>*
**Libvirt Domain Defaulting**: The action of ... *<some text here>*
**VMPreset Injection**: The action of ... *<some text here>*
.
.
.
<insert more terms>

**Comment by fabiand** [comment]
A few additional notes:

- *VirtualMachineConfig* is a static VM representation. It is called static, because it does not contain any runtime (or you could say _dynamic_)  data. The config contains a temaplte which is used to create VM instances.
- *VM* (soon eventually *VirtualMachineInstance*) represents a running Virtual Machine. It is the most detailed object when it comes to representing current runtime VM configuratoin. On the one hand it shows the current specification which was used to create the VM, on the other hand (in `.state`) it will show the actual state, including computed values like device addresses.
- *VMPreset* (not yet defined) are objects which hold partial (more or less) *VM* configurations, which are applied to *VM* up on creation. This mechanism allows a user to apply certain partial configurations to all VMs matching specific labels, this can be used to achieve something like flavours. *The exact semantics need to be defined.*
- *Expansion* (not yet defined) is the process of adding (device or attribute) dependencies to a VM definition to generate a consistent spec. This could be performed on the API level but is also performed by libvirt.
- *Defaulting* (not yet defined) is the process to decide which values to use during the expansion step. Defaulting is an inherent part of the expansion process. But we should view it as a separate step.
- *VM Template* (not yet defined) are representing a VM definition and also referencing the used resources (i.e. disks). Instantiating a *VM* from a template would lead to cloning the definition as well as the referenced resources. ([Definition from oVirt](http://www.ovirt.org/develop/release-management/features/virt/instance-types/))
- *Instance type* or *flavour* (not yet defined) define specific VM defintions (only covering the definition, not the data). An existing disk can then be started with a certain _instance type_. I.e. You got an nginx disk, which you first started with a _small_ instance type, but then switched to _huge_ because of a large amount of sessions. The disk did not change, just the vm definition. ([Definition from oVirt](http://www.ovirt.org/develop/release-management/features/virt/instance-types/).

**Comment by rmohr** [comment]
Can we move to a google sheet? I don't think it makes sense that we all start posting all our slightly different opinions here :)

https://docs.google.com/a/redhat.com/document/d/19qgN8izk9w227T9S5tRCPZvaB1Xngy718FzdFfeY3g0/edit?usp=sharing

**Comment by rmohr** [comment]
Created a public accessible doc: https://docs.google.com/document/d/12X3bwBb48GIx52g9E4DK9MiDEMH2IJKTdzxXMlG6Lng/edit?usp=sharing

**Comment by mpolednik** [comment]
Additionally, there is a great RFC that defines keywords such as MUST, SHOULD, MAY - https://www.ietf.org/rfc/rfc2119.txt. It would be helpful to follow it instead of using ambiguous "can".

**Comment by fabiand** [comment]
… the issue with the outsourced doc is that it's hard to follow any development.

**Comment by fabiand** [comment]
@rmohr @davidvossel where do we stand?

**Comment by davidvossel** [comment]
@fabiand @rmohr I'll follow up by sorting through everything we have so far and formalizing it into a patch for glossary.md. That will centralize the discussion and hopefully give us a good foundation for future (as far as definitions/terms go)
